### PR TITLE
controlplane: connect enable/disable peer to heartbeat

### PR DESCRIPTION
**Pull request by kfirtoledo**
Update peer status in policy-engine according to the heartbeat status.

### CODE REVIEW

File: pkg/controlplane/instance.go [hunk 1]
```diff
@@ -87,6 +87,15 @@ func (cp *Instance) CreatePeer(peer *cpstore.Peer) error {
 
 	cp.policyDecider.AddPeer(&api.Peer{Name: peer.Name, Spec: peer.PeerSpec})
 
+	client.SetPeerStatusCallback(func(status bool) {
```
Comment (orozery) [inline pkg/controlplane/instance.go:90]:
```suggestion
	client.SetPeerStatusCallback(func(isActive bool) {
```

File: pkg/controlplane/instance.go [hunk 2]
```diff
@@ -87,6 +87,15 @@ func (cp *Instance) CreatePeer(peer *cpstore.Peer) error {
...
+		if status {
```
Comment (orozery) [inline pkg/controlplane/instance.go:91]:
```suggestion
		if isActive {
```

File: pkg/controlplane/client.go [hunk 3]
```diff
@@ -113,11 +114,23 @@ func (c *client) IsActive() bool {
...
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		activePrevState = c.active
+		c.active = active
+		if active || c.lastSeen.IsZero() {
+			c.lastSeen = time.Now()
+		}
+	}
+	// Update other components like the policy engine with the peer status.
+	if !activePrevState && active {
```
Comment (elevran) [inline pkg/controlplane/client.go:128]:
```suggestion
	if active != activePrevState {
	   c.c.peerStatusCallback(active)
	}
```
can be an alternative to L128-134

File: pkg/controlplane/client.go [hunk 4]
```diff
@@ -113,11 +114,23 @@ func (c *client) IsActive() bool {
...
+		c.peerStatusCallback(true)
+	}
+
+	if activePrevState && !active {
+		c.peerStatusCallback(false)
 	}
```
Comment (orozery) [inline pkg/controlplane/client.go:134]:
```suggestion
	if activePrevState != active && c.peerStatusCallback != nil {
		c.peerStatusCallback(active)
	}
```

**Review by elevran — APPROVED** [review]

**Review by orozery — CHANGES_REQUESTED** [review]

File: pkg/controlplane/client.go [hunk 5]
```diff
@@ -113,11 +114,19 @@ func (c *client) IsActive() bool {
...
 func (c *client) setActive(active bool) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	c.active = active
-	if active || c.lastSeen.IsZero() {
-		c.lastSeen = time.Now()
+	var activePrevState bool
+	{
+		c.lock.Lock()
+		defer c.lock.Unlock()
```
Comment (orozery) [inline pkg/controlplane/client.go:120]: I don't think that a block helps to make this defer run when you want it.
i.e. https://stackoverflow.com/questions/49456943/why-is-gos-defer-scoped-to-function-not-lexical-enclosure
Let's revert this block, and instead call unlock directly at the end of it.

**Review by orozery — CHANGES_REQUESTED** [review]

File: pkg/controlplane/client.go [hunk 6]
```diff
@@ -113,11 +114,20 @@ func (c *client) IsActive() bool {
...
-	c.active = active
-	if active || c.lastSeen.IsZero() {
-		c.lastSeen = time.Now()
+	{
+		activePrevState = c.active
+		c.active = active
+		if active || c.lastSeen.IsZero() {
+			c.lastSeen = time.Now()
+		}
+	}
```
Comment (orozery) [inline pkg/controlplane/client.go:125]:
```suggestion
	activePrevState := c.active
	c.active = active
	if active || c.lastSeen.IsZero() {
		c.lastSeen = time.Now()
	}
```

And remove `var activePrevState bool`

**Review by orozery — CHANGES_REQUESTED** [review]

**Review by orozery — APPROVED** [review]

**Pull request merged**